token-metadata/program/src/processor/verification/collection_verification.rs

```diff
@@ -0,0 +1,218 @@
+    assert_collection_verify_is_valid(
+        &metadata.collection,
+        &collection_metadata,
+        collection_mint_info,
+        collection_master_edition_info,
+    )?;
+
+    // Determines if we have a valid authority to perform the collection verification.  The
+    // required authority is either the collection parent's metadata update authority, or a
+    // collection delegate for the collection parent.  This call fails if no valid authority is
+    // present.
+    let authority_response = AuthorityType::get_authority_type(AuthorityRequest {
+        authority: ctx.accounts.authority_info.key,
+        update_authority: &collection_metadata.update_authority,
+        mint: collection_mint_info.key,
+        metadata_delegate_record_info: ctx.accounts.delegate_record_info,
+        metadata_delegate_roles: vec![MetadataDelegateRole::Collection],
+        precedence: &[AuthorityType::Metadata, AuthorityType::MetadataDelegate],
+        ..Default::default()
+    })?;
+
+    // Validate that authority type is expected.
+    match authority_response.authority_type {
+        AuthorityType::Metadata | AuthorityType::MetadataDelegate => (),
+        _ => return Err(MetadataError::UpdateAuthorityIncorrect.into()),
+    }
+
+    // Destructure the collection field from the item metadata.
+    match metadata.collection.as_mut() {
+        Some(collection) => {
+            // Set item metadata collection to verified.
+            collection.verified = true;
+
+            // In the case of a sized collection, update the size on the collection parent.
+            if collection_metadata.collection_details.is_some() {
+                increment_collection_size(&mut collection_metadata, collection_metadata_info)?;
+            }
+        }
+        None => return Err(MetadataError::CollectionNotFound.into()),
+    };
+
+    // Reserialize metadata.
+    clean_write_metadata(&mut metadata, ctx.accounts.metadata_info)?;
+    Ok(())
+}
+
+#[allow(unreachable_code)]
+#[allow(unused_variables)]
+pub(crate) fn unverify_collection_v1(program_id: &Pubkey, ctx: Context<Unverify>) -> ProgramResult {
+    // Not tested so returning.
+    return Err(MetadataError::FeatureNotSupported.into());
+
+    // Assert program ownership/signers.
+
+    // Authority account is the collection authority and must be a signer.
+    assert_signer(ctx.accounts.authority_info)?;
+
+    if let Some(delegate_record_info) = ctx.accounts.delegate_record_info {
+        assert_owned_by(delegate_record_info, program_id)?;
+    }
+
+    assert_owned_by(ctx.accounts.metadata_info, program_id)?;
+
+    let collection_mint_info = ctx
+        .accounts
+        .collection_mint_info
+        .ok_or(MetadataError::MissingCollectionMint)?;
+    assert_owned_by(collection_mint_info, &spl_token::ID)?;
+
+    let collection_metadata_info = ctx
+        .accounts
+        .collection_metadata_info
+        .ok_or(MetadataError::MissingCollectionMetadata)?;
```

Comment:
Note to myself and everyone else: owner check for this happens later after we ensure it's not a burned collection parent, otherwise the check might fail if the burned NFT has been reassigned to the system program.

Reply:
I can add a comment here saying owner check happens later

Reply:
Added clarifying comment.